### PR TITLE
Fix plugin logging before crash

### LINKED ISSUE #3905: c-lightning crash
### Issue and Steps to Reproduce

C-lightning (v0.9.0-2-ge7d89cd) crashed while I wasn't doing anything and it was just running on my server:  C-lightning ran fine again after a restart.

Here is the relevant crash log:

```
+353275.352180714 gossipdDEBUG: Ignoring redundant update for 541045x1888x0/1 (last 1596353997, now 1596613197)
+353275.352518922 gossipdDEBUG: Received channel_update for channel 563904x1769x0/0 now ACTIVE
+353275.352913631 gossipdDEBUG: Ignoring redundant update for 563904x1769x0/0 (last 1596353997, now 1596613197)
+353275.353255043 gossipdDEBUG: Received channel_update for channel 632842x2096x0/1 now ACTIVE
+353275.353595650 gossipdDEBUG: Ignoring redundant update for 632842x2096x0/1 (last 1596281411, now 1596612611)
+353275.353648295 gossipdDEBUG: Received channel_update for channel 558820x1513x0/1 now ACTIVE
+353275.353690830 gossipdDEBUG: Ignoring redundant update for 558820x1513x0/1 (last 1596598210, now 1596612607)
+353275.353731056 gossipdDEBUG: Received channel_update for channel 600602x390x0/1 now DISABLED
+353275.353771566 gossipdDEBUG: Ignoring redundant update for 600602x390x0/1 (last 1596267449, now 1596613049)
+353275.353811632 gossipdDEBUG: Received channel_update for channel 603212x99x0/1 now ACTIVE
+353275.353866916 gossipdDEBUG: Ignoring redundant update for 603212x99x0/1 (last 1596260309, now 1596613109)
+353275.478688276 plugin-bcliIO_IN: 7b226a736f6e727063223a22322e30222c226d6574686f64223a226c6f67222c22706172616d73223a7b226c6576656c223a227761726e222c226d657373616765223a22626974636f696e2d636c6920676574626c6f636b6861736820363432333035206578697465642077697468207374617475732031227d207d0a0a
+353275.478713571 plugin-bcliUNUSUAL: bitcoin-cli getblockhash 642305 exited with status 1
+353276.639984311 plugin-bcliINFO: Killing plugin: Plugin exited before completing handshake.
+353276.640058453 lightningdBROKEN: The Bitcoin backend died.
+353277.452120031 lightningdBROKEN: FATAL SIGNAL 6 (version v0.9.0-2-ge7d89cd)
+353277.452155928 lightningdBROKEN: backtrace: common/daemon.c:43 (send_backtrace) 0x55ed94e39fc7
+353277.452169622 lightningdBROKEN: backtrace: common/daemon.c:51 (crashdump) 0x55ed94e3a012
+353277.452180698 lightningdBROKEN: backtrace: (null):0 ((null)) 0x7f98b9393f1f
+353277.452192168 lightningdBROKEN: backtrace: (null):0 ((null)) 0x7f98b9393e97
+353277.452201490 lightningdBROKEN: backtrace: (null):0 ((null)) 0x7f98b9395800
+353277.452211656 lightningdBROKEN: backtrace: lightningd/log.c:820 (fatal) 0x55ed94e1aff0
+353277.452222138 lightningdBROKEN: backtrace: lightningd/bitcoind.c:42 (bitcoin_destructor) 0x55ed94e07ae1
+353277.452232451 lightningdBROKEN: backtrace: ccan/ccan/tal/tal.c:240 (notify) 0x55ed94e88abc
+353277.452242385 lightningdBROKEN: backtrace: ccan/ccan/tal/tal.c:402 (del_tree) 0x55ed94e88b20
+353277.452271164 lightningdBROKEN: backtrace: ccan/ccan/tal/tal.c:486 (tal_free) 0x55ed94e8907d
+353277.452280448 lightningdBROKEN: backtrace: lightningd/plugin.c:217 (plugin_kill) 0x55ed94e2f7d3
+353277.452289265 lightningdBROKEN: backtrace: lightningd/plugin.c:526 (plugin_conn_finish) 0x55ed94e2f7ed
+353277.452299098 lightningdBROKEN: backtrace: ccan/ccan/io/poll.c:244 (destroy_conn) 0x55ed94e80182
+353277.452308039 lightningdBROKEN: backtrace: ccan/ccan/io/poll.c:250 (destroy_conn_close_fd) 0x55ed94e801a4
+353277.452317034 lightningdBROKEN: backtrace: ccan/ccan/tal/tal.c:240 (notify) 0x55ed94e88abc
+353277.452325753 lightningdBROKEN: backtrace: ccan/ccan/tal/tal.c:402 (del_tree) 0x55ed94e88b20
+353277.452334406 lightningdBROKEN: backtrace: ccan/ccan/tal/tal.c:486 (tal_free) 0x55ed94e8907d
+353277.452363384 lightningdBROKEN: backtrace: ccan/ccan/io/io.c:450 (io_close) 0x55ed94e7edac
+353277.452393787 lightningdBROKEN: backtrace: ccan/ccan/io/poll.c:449 (io_loop) 0x55ed94e8076c
+353277.452403540 lightningdBROKEN: backtrace: lightningd/io_loop_with_timers.c:24 (io_loop_with_timers) 0x55ed94e158aa
+353277.452413967 lightningdBROKEN: backtrace: lightningd/lightningd.c:997 (main) 0x55ed94e18f4e
+353277.452422762 lightningdBROKEN: backtrace: (null):0 ((null)) 0x7f98b9376b96
+353277.452431638 lightningdBROKEN: backtrace: (null):0 ((null)) 0x55ed94e079d9
+353277.452439506 lightningdBROKEN: backtrace: (null):0 ((null)) 0xffffffffffffffff
```

### `getinfo` output
"version": "v0.9.0-2-ge7d89cd",

### PR DESCRIPTION
**Pull request by darosior**
This aims to eventually fix #3905..

### CODE REVIEW

File: lightningd/plugin.c [hunk 1]
```diff
@@ -554,6 +554,8 @@ static struct io_plan *plugin_write_json(struct io_conn *conn,
 /* This catches the case where their stdout closes (usually they're dead). */
 static void plugin_conn_finish(struct io_conn *conn, struct plugin *plugin)
 {
+	/* Did it try to tell us something before dying ? */
+	io_flush_sync(conn);
 	plugin_kill(plugin, "Plugin exited before completing handshake.");
```
Comment (darosior) [inline lightningd/plugin.c:559]: I should also add a commit to fix this message afterwards, it's been confusing sometimes.

File: tests/test_misc.py [hunk 2]
```diff
@@ -128,6 +132,21 @@ def crash_bitcoincli(r):
     bitcoind.generate_block(5)
     sync_blockheight(bitcoind, [l1])
 
+    # When the retry timeout is reached, we should deliver a nice logging
+    # message explaining what's going on
+    l1.daemon.rpcproxy.mock_rpc('estimatesmartfee', crash_bitcoincli)
+    l1.daemon.rpcproxy.mock_rpc('getblockhash', crash_bitcoincli)
+    #fut = executor.submit(l1.rpc.call, "getrawblockbyheight", {"height": 102})
```
Comment (niftynei) [inline tests/test_misc.py:139]: delete this.

File: tests/test_misc.py [hunk 3]
```diff
@@ -128,6 +132,21 @@ def crash_bitcoincli(r):
...
+    l1.daemon.wait_for_logs(["Unable to estimate .* fee",
+                             "getblockhash .* exited with status 1",
+                             "bitcoin-cli .* getblockhash .* exited 1 .* we "
+                             "have been retrying command for --bitcoin-retry"
+                             "-timeout={} seconds with status 1"
+                             .format(retry_timeout)])
+    l1.daemon.rpcproxy.mock_rpc('estimatesmartfee', None)
+    l1.daemon.rpcproxy.mock_rpc('getblockhash', None)
+    #fut.result()
```
Comment (niftynei) [inline tests/test_misc.py:148]: and this.

**Comment by darosior** [comment]
> Didnt' realize how unfinished this is lol

Was not meant to be reviewed, but to showcase something after a discussion with @rustyrussell on IRC. That's, by the way, why it is a draft (the top commit is even called "TEMP").

> is the goal just a better log message?

Yep, the root issue of #3905 is confusion: our log messages do not arrive to `bitcoind` before the crash. I believe that fixing logging fixes the root of the issue.

-----

Closing this for now if that's confusing (someone can pick it up at will i won't be able to continue this work soon)

**Pull request closed without merging**